stumpy/scrump.py

```diff
@@ -109,36 +109,47 @@ def _compute_PI(
     for i in indices[start:stop]:
         Q = T_A[i : i + m]
         QT[:] = core._sliding_dot_product(Q, T_B)
-        # Update P[i] relative to all T[j : j + m]
         squared_distance_profile[:] = core._mass(Q, T_B, QT, μ_Q[i], σ_Q[i], M_T, Σ_T)
         squared_distance_profile[:] = np.square(squared_distance_profile)
         if excl_zone is not None:
             zone_start = max(0, i - excl_zone)
             zone_stop = min(l, i + excl_zone)
             squared_distance_profile[zone_start : zone_stop + 1] = np.inf
 
-            # only for self-join
+        if excl_zone is not None:  # self-join
+            # note: S_index = T[index: index + m]
+            # `v = squared_distance_profile[idx]` is (the square of)
+            # `dist(S_i, S_idx)`, which is the same as `dist(S_idx, S_i)`. So,
+            # `squared_distance_profile[idx]` is (the square of) distane from `S_idx`
+            # to one of its neighbors, `S_i`. Therefore, the value `v` can be
+            # used to update the TopK of `S_idx`, stored "ascendingly" in
+            # `P_squared[thread_idx, idx, :]`.
+
+            # `P_squared[thread_idx, idx, :]` in inf for those `idx` that are in
+            # the trivial zone, including the `i` itself. Those are not updated here.
             IDX = np.flatnonzero(
                 squared_distance_profile < P_squared[thread_idx, :, -1]
             )
             for idx in IDX:
+                d_squared = squared_distance_profile[idx]
                 pos = np.searchsorted(
-                    P_squared[thread_idx, idx],
-                    squared_distance_profile[idx],
-                    side="right",
+                    P_squared[thread_idx, idx], d_squared, side="right"
                 )
-                core._insert(
-                    P_squared[thread_idx, idx, :], pos, squared_distance_profile[idx]
-                )
-                core._insert(I[thread_idx, idx, :], pos, i)
+                core._shift_insert_at_index(P_squared[thread_idx, idx], pos, d_squared)
+                core._shift_insert_at_index(I[thread_idx, idx], pos, i)
 
-        idx = np.argmin(squared_distance_profile)
-        core._insert(P_squared[thread_idx, i, :], 0, squared_distance_profile[idx])
-        core._insert(I[thread_idx, i, :], 0, idx)
+        # find EXACT (not approx.) value of `P_squared[thread_idx, i, 0]`
+        nn_of_i = np.argmin(squared_distance_profile)
+        core._shift_insert_at_index(
+            P_squared[thread_idx, i], 0, squared_distance_profile[nn_of_i]
+        )
+        core._shift_insert_at_index(I[thread_idx, i], 0, nn_of_i)
 
         if P_squared[thread_idx, i, 0] == np.inf:  # pragma: no cover
             I[thread_idx, i, 0] = -1
         else:
+            # update P_squared[thread_idx, index, :] for those `index` that are
```

Comment:
Can you please elaborate? I don't understand this comment

Reply:
> Does something else need to be done?

No. I am just explaining what is going on in the following lines.  For `S_i` and its nearest neighbor `S_nn`.... we now  calculate the distances `dist(S_(i+1), S_(nn+1))`, `dist(S_(i+2), S_(nn+2))`, ... as well as `dist(S_(i-1), S_(nn-1))`, `dist(S_(i-2), S_(nn-2))`. 

We update top-k of `S_a`, and top-k of  `S_b` after calculating `dist(S_a, S_b)`. 

I think my comment was written poorly. We can just remove it if you think it is not necessary and it confuses the user/reader.

Reply:
While a comment is warranted, I do think the current wording isn't clear